navigate-route/src/main/java/com/esri/arcgismaps/sample/navigateroute/MainActivity.kt

```diff
@@ -0,0 +1,405 @@
+
+                    // listen for changes in location
+                    mapView.locationDisplay.location.collect {
+                        // get the route's tracking status
+                        val trackingStatus = routeTracker.trackingStatus.value ?: return@collect
+
+                        // displays the remaining and traversed route
+                        updateRouteGraphics(trackingStatus)
+
+                        // display route status and directions info
+                        displayRouteInfo(
+                            routeTracker,
+                            trackingStatus,
+                            simulatedLocationDataSource,
+                            routeTrackerLocationDataSource
+                        )
+                    }
+                }
+
+                launch {
+                    // if the user navigates the map view away from the
+                    // location display, activate the recenter button
+                    autoPanMode.collect {
+                        if (it == LocationDisplayAutoPanMode.Off)
+                            recenterButton.isEnabled = true
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Displays the route distance and time information using [trackingStatus],
+     * plays the voice guidance using [routeTracker]. When final destination is reached,
+     * [simulatedLocationDataSource] and [routeTrackerLocationDataSource] is stopped.
+     */
+    private fun displayRouteInfo(
+        routeTracker: RouteTracker,
+        trackingStatus: TrackingStatus,
+        simulatedLocationDataSource: SimulatedLocationDataSource,
+        routeTrackerLocationDataSource: RouteTrackerLocationDataSource,
+    ) = lifecycleScope.launch {
+
+        // get remaining distance information
+        val remainingDistance = trackingStatus.destinationProgress.remainingDistance
+        // covert remaining minutes to hours:minutes:seconds
+        val remainingTimeString = DateUtils.formatElapsedTime(
+            (trackingStatus.destinationProgress.remainingTime * 60).toLong()
+        )
+
+        // update text views
+        distanceRemainingTextView.text = getString(
+            R.string.distance_remaining, remainingDistance.displayText,
+            remainingDistance.displayTextUnits.abbreviation
+        )
+        timeRemainingTextView.text = getString(
+            R.string.time_remaining,
+            remainingTimeString
+        )
+
+        // listen for new voice guidance events
+        lifecycleScope.launch {
+            routeTracker.newVoiceGuidance.collect { voiceGuidance ->
+                // use Android's text to speech to speak the voice guidance
+                speakVoiceGuidance(voiceGuidance.text)
+                nextDirectionTextView.text = getString(
+                    R.string.next_direction,
+                    voiceGuidance.text
+                )
+            }
+        }
```

Comment:
I'm not sure about this because every time a new location is received this method is called which launches a new coroutine to collect voice guidance. But really there only needs to be one collection voice guidance in the whole application. My first thought would be to move this code into `startNavigation` although that also isn't ideal because every time the route is restarted there will be a new coroutine for collecting voice guidance. It's probably ok though.

Reply:
That's a fair point. I've updated the code so the voice guidance is called from `startNavigation` instead of every location update and moved all the voice guidance code to a fun `updateVoiceGuidance()`